README.md
# Homebridge-MagicHome-Accessory

This plugin allows MagicHome single channel devices to be switched on/off and to be dimmed in HomeKit.

## Config

```json
"platforms": [
  {
    "name": "MagicHome",
    "devices": [
      {
        "name": "Kitchen Light",
        "ip": "192.168.0.54",
        "port": 5577,
        "dimmable": true
      },
      {
        "name": "Bedroom Light",
        "ip": "192.168.0.62",
        "port": 5577,
        "dimmable": false
      }
    ],
    "connect_timeout": 1000,
    "response_timeout": 500,
    "command_timeout": 100,
    "platform": "MagicHomePlatform"
  }
],
```

### Properties

### Platform

| Property                   | Description                                                                               | Default Value            |
| -------------------------- | ----------------------------------------------------------------------------------------- | ------------------------ |
| platform *                 | The platform's name. This must not be changed!                                            | `"MagicHomePlatform"`    |
| name *                     | The name of the platform.                                                                 | `"MagicHome"`            |
| devices                    | An array of devices.                                                                      | empty                    |
| connect_timeout            | The time in `ms` after which the connection is beeing aborted without data sent/received. | `1000`                   |
| response_timeout           | The time in `ms` to wait for a response from the device.                                  | `500`                    |
| command_timeout            | The time in `ms` until a command is invalid.                                              | `100`                    |

`*` Required fields

### Devices

| Property   | Description                                                    | Default Value |
| ---------- | -------------------------------------------------------------- | ------------- |
| name *     | The name of the device that will be displayed in the Home app. | `""`          |
| ip *       | The IP addess of the device. (eg. "192.168.0.54")              | `empty`       |
| port *     | The port. Only change this if your device uses another port.   | `5577`        |
| dimmable * | Let the device be dimmable or only use an On/Off switch.       | `true`        |

`*` Required fields

## Supported Devices

At this time only single channel/color controllers are supported like this one:

![Single Channel](./.docs/single-channel.jpeg)

For RGB controllers I recommend using [homebridge-magichome2](https://www.npmjs.com/package/homebridge-magichome2).


CHANGELOG.md
# Changelog

## 1.0.3

* fixed `callback is not a function` error

## 1.0.2

* removed unneeded dependencies

## 1.0.1

* Fixed import exception

## 1.0.0

### Breaking changes

* Changed the plugin type from `accessory` to `platform`. The plugin configuration must be removed and reconfigured in `config.json`.

### Other changes

* Updated README to reflect changes.

## 0.0.8

* implemented blinking when tapping identify in `Home.app`.
* added changelog.


src/MHControl.ts
/**
 * ./src/MHControl.ts
 * @author Lukas Pistrol <lukas@pistrol.com>
 */

import net = require('net');

/** A structure containing options for the {@link MHControl} */
type MHControlOptions = {
  host: string;
  port: number;
  connect_timeout?: number;
  response_timeout?: number;
  command_timeout?: number;
};

/** A structure containing the response of an accessory */
type MHQueryResponse = {
  on: boolean;
  brightness: number;
};

/** A structure containing a Command and callbacks */
type MHCommandBlock = {
  command: Buffer;
  resolve: any;
  reject: any;
  expectReply: boolean;
};

/** A class managing the sending and receiving of data between
 * plugin and the accessory.
 */
export class MHControl {
  private _address: string;
  private _port: number;
  private _socket?: net.Socket;
  private _receivedData: Buffer;
  private _cmdQueue: MHCommandBlock[] = [];

  private _connect_timeout: number;
  private _command_timeout: number;
  private _response_timeot: number;

  private _receiveTimeout?: NodeJS.Timeout;
  private _connectTimeout?: NodeJS.Timeout;
  private _commandTimeout?: NodeJS.Timeout;
  private _preventDataSending = false;

  constructor(options: MHControlOptions) {
    this._address = options.host;
    this._port = options.port;
    this._connect_timeout = options.connect_timeout || 1000;
    this._command_timeout = options.command_timeout || 100;
    this._response_timeot = options.response_timeout || 500;
    this._receivedData = Buffer.alloc(0);
  }

  /**
   * **Receive Data**
   *
   * Query the current state of the device
   *
   * @param  {boolean} empty
   * @param  {Buffer} data?
   * @returns void
   */
  _receiveData(empty: boolean, data?: Buffer): void {
    if (this._commandTimeout !== undefined) {
      clearTimeout(this._commandTimeout);
      this._commandTimeout = undefined;
    }

    if (empty) {
      const finished_command = this._cmdQueue[0];

      if (finished_command !== undefined) {
        const resolve = finished_command.resolve;
        if (resolve !== undefined) {
          resolve(this._receivedData);
        }
      }

      this._receivedData = Buffer.alloc(0);

      this._cmdQueue.shift();

      this._handleNextCommand();
    } else if (data !== undefined) {
      this._receivedData = Buffer.concat([this._receivedData, data]);

      if (this._receiveTimeout !== undefined) {
        clearTimeout(this._receiveTimeout);
      }

      this._receiveTimeout = setTimeout(() => {
        this._receiveData(true);
      }, this._response_timeot);
    }
  }

  /**
   * **Handle Command Timeout**
   *
   * This method is called when the device does not respond to a command within the specified timeout.
   */
  _handleCommandTimeout() {
    this._commandTimeout = undefined;

    const timeout_command = this._cmdQueue[0];

    if (timeout_command !== undefined) {
      const reject = timeout_command.reject;
      if (reject !== undefined) {
        reject(new Error('Command timeout'));
      }
    }

    this._receivedData = Buffer.alloc(0);

    this._cmdQueue.shift();

    this._handleNextCommand();
  }

  /**
   * **Handle Next Command**
   *
   * This method is called when the device has finished processing the previous command.
   */
  _handleNextCommand() {
    if (this._cmdQueue.length === 0) {
      if (this._socket !== null) {
        this._socket?.end();
      }
      this._socket = undefined;
    } else {
      const cmd = this._cmdQueue[0];

      if (!cmd.expectReply) {
        this._socket?.write(cmd.command, 'binary', () => {
          this._receiveData(true);
        });
      } else {
        this._socket?.write(cmd.command, 'binary', () => {
          if (this._command_timeout === undefined) {
            return;
          }
          this._commandTimeout = setTimeout(() => {
            this._handleCommandTimeout();
          }, this._command_timeout);
        });
      }
    }
  }

  /**
   * **Send Command**
   *
   * This method is called when a command is sent to the device.
   *
   * @param  {Buffer} buffer
   * @param  {boolean} expectReply
   * @param  {any} resolve
   * @param  {any} reject
   * @returns void
   */
  _sendCommand(buffer: Buffer, expectReply: boolean, resolve: any, reject: any): void {
    let checksum = 0;
    for (const byte of buffer.values()) {
      checksum += byte;
    }
    checksum &= 0xff;

    const command = Buffer.concat([buffer, Buffer.from([checksum])]);

    if (this._cmdQueue.length === 0 && this._socket === undefined) {
      this._cmdQueue.push({ command, resolve, reject, expectReply });

      this._preventDataSending = false;

      this._socket = net.connect(this._port, this._address, () => {
        if (this._connectTimeout !== undefined) {
          clearTimeout(this._connectTimeout);
          this._connectTimeout = undefined;
        }
        if (!this._preventDataSending) {
          this._socket && this._socket.write(command, 'binary');
        }
      });

      this._socket.on('error', (error) => {
        this._socket = undefined;
        reject(error);
      });

      this._socket.on('data', (data) => {
        this._receiveData(false, data);
      });

      if (this._connectTimeout !== undefined) {
        this._connectTimeout = setTimeout(() => {
          this._socket = undefined;
          reject(new Error('Connection timeout'));
        }, this._connect_timeout);
      }
    } else {
      this._cmdQueue.push({ command, resolve, reject, expectReply });
    }
  }

  /**
   * **Socket Error Handler**
   *
   * This method is called when an error occurs on the socket.
   *
   * @param  {Error} err
   * @param  {any} reject
   */
  _socketErrorHandler(err: Error, reject: any) {
    this._preventDataSending = true;

    reject(err);

    if (this._socket !== undefined) {
      this._socket.end();
    }
    this._socket = undefined;

    for (const c of this._cmdQueue) {
      const reject = c.reject;
      if (reject !== undefined) {
        reject(err);
      }
    }

    this._cmdQueue = [];
  }

  /**
   * **Send Brightness Command**
   *
   * This method is called when sending the brightness to the device.
   *
   * @param  {number} brightness
   * @param  {any} callback
   * @returns {Promise<boolean>} A Promise
   */
  sendBrightnessCommand(brightness: number, callback?: any): Promise<boolean> {
    const cmd_buf = Buffer.from([0x31, brightness, 0, 0, 0x03, 0x01, 0x0f]);

    const promise = new Promise<any>((resolve, reject) => {
      this._sendCommand(cmd_buf, false, resolve, reject);
    })
      .then((data) => {
        return data.count > 0;
      })
      .then((result) => {
        return result;
      });

    if (callback && typeof callback === 'function') {
      promise.then(callback.bind(null, null), callback());
    }

    return promise;
  }

  /**
   * **Send Power Command**
   *
   * This method is called when sending the power state to the device.
   *
   * @param  {boolean} on
   * @param  {any} callback?
   * @returns {Promise<boolean>} A Promise
   */
  setPower(on: boolean, callback?: any): Promise<boolean> {
    const cmd_buf = Buffer.from([0x71, on ? 0x23 : 0x24, 0x0f]);

    const promise = new Promise<any>((resolve, reject) => {
      this._sendCommand(cmd_buf, false, resolve, reject);
    }).then((data) => {
      return data.lenght > 0;
    });

    if (callback && typeof callback === 'function') {
      promise.then(callback.bind(null, null), callback());
    }

    return promise;
  }

  /**
   * **Query State**
   *
   * This method is called when querying the state of the device.
   *
   * @param  {any} callback?
   * @returns {Promise<boolean>} A Promise
   */
  queryState(callback?: any): Promise<MHQueryResponse> {
    const cmd_buf = Buffer.from([0x81, 0x8a, 0x8b]);

    const promise = new Promise<any>((resolve, reject) => {
      this._sendCommand(cmd_buf, true, resolve, reject);
    })
      .then((data) => {
        if (data.length !== 14) {
          throw new Error('Invalid response');
        }

        const state: MHQueryResponse = {
          on: data.readUInt8(2) === 0x23,
          brightness: data.readUInt8(6),
        };

        return state;
      })
      .catch((err) => {
        callback(err);
        return err;
      });

    if (callback && typeof callback === 'function') {
      promise.then(callback.bind(null, null), callback());
    }

    return promise;
  }
}


src/accessory.ts
import {
  AccessoryConfig,
  AccessoryPlugin,
  API,
  CharacteristicEventTypes,
  CharacteristicGetCallback,
  CharacteristicSetCallback,
  Characteristic,
  CharacteristicValue,
  Logger,
  Service,
} from 'homebridge';
import { MHControl } from './MHControl';
import './extensions';

export class MagicHomeAccessory implements AccessoryPlugin {
  public readonly Service: typeof Service = this.api.hap.Service;
  public readonly Characteristic: typeof Characteristic = this.api.hap.Characteristic;
  MHOnControl: MHControl;
  MHBrightnessControl: MHControl;

  constructor(public readonly log: Logger, public readonly config: AccessoryConfig, public readonly api: API) {
    this.log = log;
    this.config = config;
    this.api = api;

    const configuration = {
      host: this.config.ip,
      port: this.config.port || 5577,
      connect_timeout: this.config.connect_timeout || 1000,
      response_timeout: this.config.response_timeout || 500,
      command_timeout: this.config.command_timeout || 100,
    };

    this.MHOnControl = new MHControl(configuration);
    this.MHBrightnessControl = new MHControl(configuration);

    this.log.debug('MagicHomePlugin finished initializing!');
    this.log.debug('IP: ' + this.config.ip);
    this.log.debug('Timeout: ' + this.config.timeout);
    this.log.debug('Apply Masks: ' + this.config.apply_masks);
  }

  identify(): void {
    this.log.info('Identify!');
  }

  getServices(): Service[] {
    const informationService = new this.Service.AccessoryInformation();

    informationService
      .setCharacteristic(this.Characteristic.Manufacturer, 'MagicHome')
      .setCharacteristic(this.Characteristic.Model, 'LED Controller')
      .setCharacteristic(this.Characteristic.SerialNumber, '123-456-789')
      .setCharacteristic(this.Characteristic.FirmwareRevision, '0.0.1');

    const lightbulbService = new this.Service.Lightbulb(this.config.name);

    lightbulbService
      .getCharacteristic(this.Characteristic.On)
      .on(CharacteristicEventTypes.GET, this.getPowerState.bind(this))
      .on(CharacteristicEventTypes.SET, this.setPowerState.bind(this));

    lightbulbService
      .addCharacteristic(this.Characteristic.Brightness)
      .on(CharacteristicEventTypes.GET, this.getBrightness.bind(this))
      .on(CharacteristicEventTypes.SET, this.setBrightness.bind(this));

    return [informationService, lightbulbService];
  }

  getPowerState(callback: CharacteristicGetCallback) {
    this.MHOnControl.queryState()
      .then((state) => {
        const isOn = this.powerState(state.on);
        this.log.debug('Power: "' + isOn + '"');
        callback(null, state.on);
      })
      .catch((error) => {
        this.log.error('Error getting power state: ' + error);
        callback(error);
      });
  }

  setPowerState(value: CharacteristicValue, callback: CharacteristicSetCallback) {
    this.MHOnControl.setPower(value as boolean)
      .then(() => {
        this.log.debug('Power Set: "' + this.powerState(value as boolean) + '"');
        callback(null);
      })
      .catch((error) => {
        this.log.error('Error setting power state: ' + error);
        callback(error);
      });
  }

  getBrightness(callback: CharacteristicGetCallback) {
    this.MHBrightnessControl.queryState()
      .then((state) => {
        const brightness = state.brightness.convertToPercent(255);
        this.log.debug('Brightness: "' + brightness + '%"');
        callback(null, brightness);
      })
      .catch((error) => {
        this.log.error('Error getting brightness: ' + error);
        callback(error);
      });
  }

  setBrightness(value: CharacteristicValue, callback: CharacteristicSetCallback) {
    this.MHBrightnessControl.sendBrightnessCommand((value as number).convertToColor(255))
      .then(() => {
        this.log.debug('Brightness Set: "' + value + '%"');
        callback(null);
      })
      .catch((error) => {
        this.log.error('Error setting brightness: ' + error);
        callback(error);
      });
  }

  private powerState(state: boolean): string {
    return state ? 'On' : 'Off';
  }
}


src/extensions.ts
/**
 * ./src/extensions.ts
 * @author Lukas Pistrol <lukas@pistrol.com>
 */

// eslint-disable-next-line @typescript-eslint/no-unused-vars
interface Number {
  /**
   * Converts a number [0-max] to a percentage value [0-100].
   * @param  {number} max The maximum value of the number.
   * @returns {number} The percentage value.
   */
  convertToPercent(max: number): number;

  /**
   * Converts a percentage value [0-100] to a number [0-max].
   * @param max The maximum value of the number.
   * @returns {number} The number.
   */
  convertToColor(max: number): number;

  /**
   * Clamps a number between a minimum and maximum value.
   * @param min The minimum value of the number.
   * @param max The maximum value of the number.
   * @returns {number} The clamped number.
   */
  clamp(min: number, max: number): number;
}

Number.prototype.convertToPercent = function (max: number) {
  return Math.round((Number(this) / max) * 100).clamp(0, 100);
};

Number.prototype.convertToColor = function (max: number) {
  return Math.round((Number(this) / 100) * max).clamp(0, max);
};

Number.prototype.clamp = function (min: number, max: number) {
  return Math.min(Math.max(Number(this), min), max);
};

// eslint-disable-next-line @typescript-eslint/no-unused-vars
interface Boolean {
  /**
   * Converts a boolean to a string of ['On', 'Off']
   * @returns string
   */
  onOff(): string;
}

Boolean.prototype.onOff = function () {
  return this ? 'On' : 'Off';
};

/**
 * A promise that resolves after a given time in ms.
 * @param  {number} ms
 */
// eslint-disable-next-line @typescript-eslint/no-unused-vars
function delay(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}


src/MagicHomePlatform.ts
/**
 * ./src/MagicHomePlatform.ts
 * @author Lukas Pistrol <lukas@pistrol.com>
 */

import {
  API,
  DynamicPlatformPlugin,
  Logger,
  PlatformAccessory,
  PlatformConfig,
  Service,
  Characteristic,
} from 'homebridge';

import { PLATFORM_NAME, PLUGIN_NAME } from './settings';
import { MHPlatformAccessory } from './MHPlatformAccessory';

/**
 * This is a type for the device configuration.
 */
export type Device = {
  name: string;
  ip: string;
  port: number;
  dimmable: boolean;
};

function isEqual(a: Device, b: Device): boolean {
  return a.ip === b.ip && a.port === b.port && a.name === b.name && a.dimmable === b.dimmable;
}

export class MagicHomePlatform implements DynamicPlatformPlugin {
  public readonly Service: typeof Service = this.api.hap.Service;
  public readonly Characteristic: typeof Characteristic = this.api.hap.Characteristic;

  public readonly accessories: PlatformAccessory[] = [];

  constructor(public readonly log: Logger, public readonly config: PlatformConfig, public readonly api: API) {
    this.log = log;
    this.config = config;
    this.api = api;

    this.log.debug('Finished initializing platform:', this.config.name);

    this.api.on('didFinishLaunching', () => {
      this.log.debug('Executed didFinishLaunching callback');
      this.discoverDevices();
    });
  }

  /**
   * This function is invoked when homebridge restores cached accessories from disk at startup.
   * It should be used to setup event handlers for characteristics and update respective values.
   */
  configureAccessory(accessory: PlatformAccessory) {
    this.log.info('Loading accessory from cache:', accessory.displayName);
    this.api.updatePlatformAccessories([accessory]);
    this.accessories.push(accessory);
  }

  /** This funtion will be called after `didFinishLaunching` event. It gets the configured
   * devices from the config.json and adds them to the homebridge platform. If a device is already
   * cached it will be restored from cache. Should a device have been changed in the config.json
   * it will be removed and added again.
   */
  discoverDevices() {
    this.log.debug('Discovering devices');

    this.config.devices.forEach((device: Device) => {
      this.log.debug('Device found:', device.name);

      const uuid = this.api.hap.uuid.generate(device.ip);

      const existingAccessory = this.accessories.find((accessory) => accessory.UUID === uuid);

      if (existingAccessory) {
        if (!isEqual(existingAccessory.context.device, device)) {
          this.log.warn('Device configuration changed, updating accessory', existingAccessory.displayName);
          this.api.unregisterPlatformAccessories(PLUGIN_NAME, PLATFORM_NAME, [existingAccessory]);
          this._addDevice(device, uuid);
        } else {
          this.log.info('Restoring existing accessory from cache:', existingAccessory.displayName);
          new MHPlatformAccessory(this, existingAccessory, device);
        }
      } else {
        this._addDevice(device, uuid);
      }
    });
  }

  /** Adds a device to the platform */
  _addDevice(device: any, uuid: string) {
    this.log.info('Adding new accessory:', device.name);

    const accessory = new this.api.platformAccessory(device.name, uuid);
    accessory.context.device = device;

    new MHPlatformAccessory(this, accessory, device);

    this.api.registerPlatformAccessories(PLUGIN_NAME, PLATFORM_NAME, [accessory]);
  }
}


src/MHPlatformAccessory.ts
/**
 * ./src/MHPlatformAccessory.ts
 * @author Lukas Pistrol <lukas@pistrol.com>
 */

import {
  Service,
  PlatformAccessory,
  CharacteristicValue,
  CharacteristicGetCallback,
  CharacteristicSetCallback,
} from 'homebridge';

import { MagicHomePlatform } from './MagicHomePlatform';

import { ACCESSORY_NAME, MANUFACTURER } from './settings';
import { version as VERSION } from '../package.json';

import { MHControl } from './MHControl';
import { Device } from './MagicHomePlatform';
import './extensions';

/**
 * This is the accessory class that is used to create the accessory in homebridge.
 */
export class MHPlatformAccessory {
  private service: Service;

  private MHOnControl: MHControl;
  private MHBrightnessControl: MHControl;

  /**
   * @param platform The platform that this accessory belongs to.
   * @param accessory The accessory that this class is wrapping.
   * @param device The device configuration for this accessory.
   */
  constructor(
    private readonly platform: MagicHomePlatform,
    private readonly accessory: PlatformAccessory,
    private readonly device: Device,
  ) {
    // set accessory configuration
    const configuration = {
      host: this.device.ip,
      port: this.device.port,
      connect_timeout: this.platform.config.connect_timeout,
      response_timeout: this.platform.config.response_timeout,
      command_timeout: this.platform.config.command_timeout,
    };

    // set controllers for power and brightness
    this.MHOnControl = new MHControl(configuration);
    this.MHBrightnessControl = new MHControl(configuration);

    // set accessory information
    this.accessory
      .getService(this.platform.Service.AccessoryInformation)!
      .setCharacteristic(this.platform.Characteristic.Manufacturer, MANUFACTURER)
      .setCharacteristic(this.platform.Characteristic.Model, ACCESSORY_NAME)
      .setCharacteristic(this.platform.Characteristic.SerialNumber, this.device.ip)
      .setCharacteristic(this.platform.Characteristic.FirmwareRevision, VERSION)
      .setCharacteristic(this.platform.Characteristic.Identify, true);

    // get the LightBulb service if it exists, otherwise create a new LightBulb service
    this.service =
      this.accessory.getService(this.platform.Service.Lightbulb) ||
      this.accessory.addService(this.platform.Service.Lightbulb);

    // set the service name, this is what is displayed as the default name on the Home app
    this.service.setCharacteristic(this.platform.Characteristic.Name, this.device.name);

    // register handlers for the On/Off Characteristic
    this.service
      .getCharacteristic(this.platform.Characteristic.On)
      .on('get', this.getPowerState.bind(this))
      .on('set', this.setPowerState.bind(this));

    // register handlers for the Brightness Characteristic if the device is dimmable
    if (this.device.dimmable) {
      this.service
        .getCharacteristic(this.platform.Characteristic.Brightness)
        .on('get', this.getBrightness.bind(this))
        .on('set', this.setBrightness.bind(this));
    }
  }

  /**
   * **Get Power State**
   *
   * This method gets the power state of the lightbulb.
   *
   * @param  {CharacteristicGetCallback} callback
   */
  getPowerState(callback: CharacteristicGetCallback) {
    this.MHOnControl.queryState()
      .then((state) => {
        const isOn = state.on.onOff();
        this.platform.log.debug(`[${this.device.name}] Power: "${isOn}"`);
        callback(null, state.on);
      })
      .catch((error) => {
        this.platform.log.error(`[${this.device.name}] Error getting power state:`, error);
        callback(error);
      });
  }

  /**
   * **Set Power State**
   *
   * This method sets the power state of the lightbulb.
   *
   * @param  {CharacteristicValue} value
   * @param  {CharacteristicSetCallback} callback
   */
  setPowerState(value: CharacteristicValue, callback: CharacteristicSetCallback) {
    this.MHOnControl.setPower(value as boolean)
      .then(() => {
        this.platform.log.info(`[${this.device.name}] Power Set: "${(value as boolean).onOff()}"`);
        callback(null);
      })
      .catch((error) => {
        this.platform.log.error(`[${this.device.name}] Error setting power state:`, error);
        callback(error);
      });
  }

  /**
   * **Get Brightness**
   *
   * This method gets the brightness of the lightbulb.
   *
   * @param  {CharacteristicGetCallback} callback
   */
  getBrightness(callback: CharacteristicGetCallback) {
    this.MHBrightnessControl.queryState()
      .then((state) => {
        const brightness = state.brightness.convertToPercent(255);
        this.platform.log.debug(`[${this.device.name}] Brightness: "${brightness}%"`);
        callback(null, brightness);
      })
      .catch((error) => {
        this.platform.log.error(`[${this.device.name}] Error getting brightness:`, error);
        callback(error);
      });
  }

  /**
   * **Set Brightness**
   *
   * This method sets the brightness of the lightbulb.
   *
   * @param  {CharacteristicValue} value
   * @param  {CharacteristicSetCallback} callback
   */
  setBrightness(value: CharacteristicValue, callback: CharacteristicSetCallback) {
    this.MHBrightnessControl.sendBrightnessCommand((value as number).convertToColor(255))
      .then(() => {
        this.platform.log.info(`[${this.device.name}] Brightness Set: "${value}%"`);
        callback(null);
      })
      .catch((error) => {
        this.platform.log.error(`[${this.device.name}] Error setting brightness:`, error);
        callback(error);
      });
  }
}


src/MagicHomeAccessory.ts
/**
 * ./src/MagicHomeAccessory.ts
 * @author Lukas Pistrol <lukas@pistrol.com>
 */

import {
  AccessoryConfig,
  AccessoryPlugin,
  API,
  CharacteristicEventTypes,
  CharacteristicGetCallback,
  CharacteristicSetCallback,
  Characteristic,
  CharacteristicValue,
  Logger,
  Service,
} from 'homebridge';
import { MHControl } from './MHControl';
import { ACCESSORY_NAME, MANUFACTURER } from './settings';
import { version as VERSION } from '../package.json';
import './extensions';

/** A class representing the accessory plugin itself.
 * It implements {@link AccessoryPlugin}.
 */
export class MagicHomeAccessory implements AccessoryPlugin {
  public readonly Service: typeof Service = this.api.hap.Service;
  public readonly Characteristic: typeof Characteristic = this.api.hap.Characteristic;
  MHOnControl: MHControl;
  MHBrightnessControl: MHControl;

  constructor(public readonly log: Logger, public readonly config: AccessoryConfig, public readonly api: API) {
    this.log = log;
    this.config = config;
    this.api = api;

    const configuration = {
      host: this.config.ip,
      port: this.config.port || 5577,
      connect_timeout: this.config.connect_timeout || 1000,
      response_timeout: this.config.response_timeout || 500,
      command_timeout: this.config.command_timeout || 100,
    };

    this.MHOnControl = new MHControl(configuration);
    this.MHBrightnessControl = new MHControl(configuration);

    this.log.debug('MagicHomePlugin finished initializing: ' + this.config.name);
    this.log.debug('IP: ' + this.config.ip);
    this.log.debug('Configuration: ' + JSON.stringify(configuration));
  }

  /**
   * **Identify Accessory**
   *
   * This method is called when tapping the "identify" button in the Home app during setup.
   * The accessory will turn on for 0.5 seconds and then turn off.
   *
   * @returns void
   */
  identify(): void {
    this.log.info('Identify!');
    this.MHBrightnessControl.setPower(true).then(() => {
      delay(500).then(() => {
        this.MHOnControl.setPower(false);
      });
    });
  }

  /**
   * **Get Services**
   *
   * This method sets up the accessory's services and characteristics.
   * It is called once after the accessory is initialized.
   *
   * @returns {Service[]} The services of the accessory
   */
  getServices(): Service[] {
    const informationService = new this.Service.AccessoryInformation();

    informationService
      .setCharacteristic(this.Characteristic.Manufacturer, MANUFACTURER)
      .setCharacteristic(this.Characteristic.Model, ACCESSORY_NAME)
      .setCharacteristic(this.Characteristic.SerialNumber, this.config.ip)
      .setCharacteristic(this.Characteristic.FirmwareRevision, VERSION)
      .setCharacteristic(this.Characteristic.Identify, true);

    const lightbulbService = new this.Service.Lightbulb(this.config.name);
    const enabledCharacteristics = this.config.enabled_characteristics || [];

    this.log.debug('Enabled Characteristics: ' + enabledCharacteristics);

    if (enabledCharacteristics.includes('On')) {
      lightbulbService
        .getCharacteristic(this.Characteristic.On)
        .on(CharacteristicEventTypes.GET, this.getPowerState.bind(this))
        .on(CharacteristicEventTypes.SET, this.setPowerState.bind(this));
    }

    if (enabledCharacteristics.includes('Brightness')) {
      lightbulbService
        .addCharacteristic(this.Characteristic.Brightness)
        .on(CharacteristicEventTypes.GET, this.getBrightness.bind(this))
        .on(CharacteristicEventTypes.SET, this.setBrightness.bind(this));
    }

    return [informationService, lightbulbService];
  }

  /**
   * **Get Power State**
   *
   * This method gets the power state of the lightbulb.
   *
   * @param  {CharacteristicGetCallback} callback
   */
  getPowerState(callback: CharacteristicGetCallback) {
    this.MHOnControl.queryState()
      .then((state) => {
        const isOn = state.on.onOff();
        this.log.debug('Power: "' + isOn + '"');
        callback(null, state.on);
      })
      .catch((error) => {
        this.log.error('Error getting power state: ' + error);
        callback(error);
      });
  }

  /**
   * **Set Power State**
   *
   * This method sets the power state of the lightbulb.
   *
   * @param  {CharacteristicValue} value
   * @param  {CharacteristicSetCallback} callback
   */
  setPowerState(value: CharacteristicValue, callback: CharacteristicSetCallback) {
    this.MHOnControl.setPower(value as boolean)
      .then(() => {
        this.log.info('Power Set: "' + (value as boolean).onOff() + '"');
        callback(null);
      })
      .catch((error) => {
        this.log.error('Error setting power state: ' + error);
        callback(error);
      });
  }

  /**
   * **Get Brightness**
   *
   * This method gets the brightness of the lightbulb.
   *
   * @param  {CharacteristicGetCallback} callback
   */
  getBrightness(callback: CharacteristicGetCallback) {
    this.MHBrightnessControl.queryState()
      .then((state) => {
        const brightness = state.brightness.convertToPercent(255);
        this.log.debug('Brightness: "' + brightness + '%"');
        callback(null, brightness);
      })
      .catch((error) => {
        this.log.error('Error getting brightness: ' + error);
        callback(error);
      });
  }

  /**
   * **Set Brightness**
   *
   * This method sets the brightness of the lightbulb.
   *
   * @param  {CharacteristicValue} value
   * @param  {CharacteristicSetCallback} callback
   */
  setBrightness(value: CharacteristicValue, callback: CharacteristicSetCallback) {
    this.MHBrightnessControl.sendBrightnessCommand((value as number).convertToColor(255))
      .then(() => {
        this.log.info('Brightness Set: "' + value + '%"');
        callback(null);
      })
      .catch((error) => {
        this.log.error('Error setting brightness: ' + error);
        callback(error);
      });
  }
}
